approxFutureEnergy.m

```diff
@@ -1,131 +1,176 @@
+
+elseif (eta < 0)
+  [W2] = icare(A, B, (C.' * C), R, 'anti');
+
+  if (isempty(W2) && verbose)
+    warning('approxFutureEnergy: icare couldn''t find stabilizing solution')
+    warning('approxFutureEnergy: using the hamiltonian')
+    [~, W2, ~] = hamiltonian(A, B, C.' * C, R, true);
   end
 
-  %  Reshape the resulting quadratic coefficients
-  w2 = W2(:);
-  w{2} = w2;
+else % eta==0
+  [W2] = lyap(A.', (C.' * C));
 
-  %% k=3 case
-  if ( d>2 ) % set up the generalized Lyapunov solver
-    W2BB = W2*(B*B.');
-    Acell = cell(1,d);
-    for i=1:d
-      Acell{i} = A.'-eta*W2BB;
-    end
+end
+
+if (isempty(W2))
+  error('approxFutureEnergy: Can''t find a stabilizing solution')
+end
+
+%  Check the residual of the Riccati/Lyapunov equation
+if (verbose)
+  RES = A' * W2 + W2 * A - eta * (W2 * B) * (B' * W2) + C' * C;
+  fprintf('The residual of the Riccati equation is %g\n', norm(RES, 'inf'));
+end
+
+%  Reshape the resulting quadratic coefficients
+w{2} = vec(W2);
+
+%% k=3 case
+if (d > 2)
+  GaWb = cell(2 * l + 1, d - 1); % Pre-compute G_a.'*W_b, etc for all the a,b we need
+  GaWb{1, 2} = B.' * W2;
+  Im = speye(m);
+  % set up the generalized Lyapunov solver
+  [Acell{1:d}] = deal(A.' - eta * W2 * (B * B.'));
 
-    b = -LyapProduct(N.',w2,2);
-    [w{3}] = KroneckerSumSolver(Acell(1:3),b,3);
-    
-    [w{3}] = kronMonomialSymmetrize(w{3},n,3);
+  b = -LyapProduct(N.', w{2}, 2);
+
+  if l > 0 % New for QB/polynomial input
+    GaWb{2, 2} = g{2}.' * W2;
+    b = b + 2 * eta * kron(speye(n ^ 3), vec(Im).') * vec(kron(GaWb{2, 2}, GaWb{1, 2}));
   end
-  
+
+  [w{3}] = KroneckerSumSolver(Acell(1:3), b, 3); % Solve Ax=b for k=3
+  [w{3}] = kronMonomialSymmetrize(w{3}, n, 3); % Symmetrize w3
+
   %% k>3 cases (up to d)
-  BWk = cell(1,d-1);
-  if ( d>3 )
-
-    for k=4:d
-      BWk{k-1} = B.'*reshape(w{k-1},n,n^(k-2));
-
-      b = -LyapProduct(N.',w{k-1},k-1);
-      
-      for i=3:(k+1)/2
-        j   = k+2-i;
-        tmp = BWk{i}.'*BWk{j};
-        b   = b + 0.25*eta*i*j*(vec(tmp) + vec(tmp.'));
-      end
+  for k = 4:d
+    GaWb{1, k - 1} = B.' * reshape(w{k - 1}, n, n ^ (k - 2));
 
-      if (mod(k,2)==0) % k is even
-        i   = (k+2)/2; 
-        j   = i;
-        tmp = BWk{i}.'*BWk{j};
-        b   = b + 0.25*eta*i*j*vec(tmp);
-      end
+    b = -LyapProduct(N.', w{k - 1}, k - 1); % Pre-compue all the L(N') terms
 
-      [w{k}] = KroneckerSumSolver(Acell(1:k),b,k);
+    % Now add all the terms from the 'B' sum by looping through the i and j
+    for i = 3:(k + 1) / 2 % i+j=k+2
+      j = k + 2 - i;
+      tmp = GaWb{1, i}.' * GaWb{1, j};
+      b = b + 0.25 * eta * i * j * (vec(tmp) + vec(tmp.'));
+    end
 
-      [w{k}] = kronMonomialSymmetrize(w{k},n,k);
+    if ~mod(k, 2) % k is even
+      i = (k + 2) / 2;
+      j = i;
+      tmp = GaWb{1, i}.' * GaWb{1, j};
+      b = b + 0.25 * eta * i * j * vec(tmp);
     end
+
+    % Now add the higher order polynomial terms by iterating through the sums
```

Comment:
This block (lines 148-168) is a generalization of lines 120-123 but for general w{k}.